### DISCUSSION TITLE
Knitro which_depends error

### DISCUSSION THREAD

**Discussion by tinusd** [casadi/casadi#3248] [Q&A] [answered]
Hi there

I am using C-code generation for my NLP. I am using Matlab and Knitro on Windows. I do the following:
```
solver = casadi.nlpsol('solver', 'knitro', nlp);
solver.generate_dependencies('prob.c');
system('cl.exe prob.c /Ox /link /DLL /OUT:prob.dll')
solver2 = casadi.nlpsol('solver', 'knitro', 'prob.dll');
```
and I get the error:
Error using casadi.nlpsol (line 1639)
.../casadi/core/function_internal.cpp:145: Error calling KnitroInterface::init for 'solver':
Error in Function::which_depends for 'nlp' [External] at .../casadi/core/function.cpp:1769:
.../casadi/core/function_internal.cpp:3558: 'which_depends' not defined for External

Error in test_problems (line 44)
solver2 = casadi.nlpsol('solver', 'knitro', 'prob.dll');

If I use IPOPT as my solver, everything goes smooth. Any idea what the problem is?

Thanks a lot in advance!

**Comment by tinusd** [comment]
I have not been able to solve the problem. It seems like the problem is only when I try to use Knitro When I use the sqpmethod (or IPOPT), I get no error. Could this be a bug/issue? Or I am forgetting something?

**Comment by tinusd** [comment]
I managed to track down the problem, I believe.

When I looked into knitro_interface.cpp, I saw the following lines:
```
    // Type of constraints, general by default
    if (contype_.empty()) {
      contype_.resize(ng_, KN_CONTYPE_GENERAL);
      if (detect_linear_constraints) {
        std::vector<bool> nl_g = oracle_.which_depends("x", {"g"}, 2, true);
        for (casadi_int i=0;i<ng_;++i)
          contype_[i] = nl_g[i] ? KN_CONTYPE_GENERAL : KN_CONTYPE_LINEAR;
      }
    }
```
I noticed the which_depends. I bypassed it by setting the option 
`option.detect_linear_constraints = false`,
which I pass to nlpsol. I can then load the dll without any problems.

Should I open an issue for this?

**Reply by jgillis** [reply, marked as answer]
This is indeed an omission on our part, thanks for reporting. Spawned #3257
